gateway/handler_defaults.go

```diff
@@ -0,0 +1,207 @@
+package gateway
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ipfs/go-libipfs/files"
+	mc "github.com/multiformats/go-multicodec"
+
+	ipath "github.com/ipfs/interface-go-ipfs-core/path"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+)
+
+func (i *handler) serveDefaults(ctx context.Context, w http.ResponseWriter, r *http.Request, maybeResolvedImPath ImmutablePath, immutableContentPath ImmutablePath, contentPath ipath.Path, begin time.Time, requestedContentType string, logger *zap.SugaredLogger) bool {
+	ctx, span := spanTrace(ctx, "ServeDefaults", trace.WithAttributes(attribute.String("path", contentPath.String())))
+	defer span.End()
+
+	var (
+		pathMetadata           ContentPathMetadata
+		bytesResponse          files.File
+		isDirectoryHeadRequest bool
+		directoryMetadata      *directoryMetadata
+		err                    error
+	)
+
+	switch r.Method {
+	case http.MethodHead:
+		var data files.Node
+		pathMetadata, data, err = i.api.Head(ctx, maybeResolvedImPath)
+		if !i.handleRequestErrors(w, contentPath, err) {
+			return false
+		}
+		defer data.Close()
+		if _, ok := data.(files.Directory); ok {
+			isDirectoryHeadRequest = true
+		} else if f, ok := data.(files.File); ok {
+			bytesResponse = f
+		} else {
+			webError(w, fmt.Errorf("unsupported response type"), http.StatusInternalServerError)
+			return false
+		}
+	case http.MethodGet:
+		// TODO: refactor below: we should not have 2x20 duplicated flow control when the only difference is ranges.
+		rangeHeader := r.Header.Get("Range")
+		if rangeHeader == "" {
+			var getResp *GetResponse
+			// TODO: passing resolved path here, instead of contentPath is harming content routing. Knowing original immutableContentPath will allow backend to find  providers for parents, even when internal CIDs are not announced, and will provide better key for caching related DAGs.
+			pathMetadata, getResp, err = i.api.Get(ctx, maybeResolvedImPath)
```

Comment:
:warning:  @aschmahmann  how bad would it be to just remove `maybeResolvedImPath`? 
Resolution results are highly cacheable, it does not really matter if we do it once or twice in one system, but it potentially closes the door for content routing optimization here, makes code more complex, and I suspect it also opens a surface for bugs like the one described in TODO about  `X-Ipfs-Roots` below.

Reply:
Ack, let's park this until https://github.com/ipfs/boxo/issues/221, resolving for now.